### Introduction

In [1]:
import pandas as pd
import numpy as np

# read the data set into a pandas dataframe
f500 = pd.read_csv("f500.csv", index_col=0)
f500.index.name = None

# replace 0 values in the "previous_rank" column with NaN
f500.loc[f500["previous_rank"] == 0, "previous_rank"] = np.nan

1. Select the `rank`, `revenues`, and `revenue_change` columns in `f500`. Then, use the `DataFrame.head()` method to select the first 5 rows. Assign the result to `f500_selection`.

In [2]:
f500_selection = f500[["rank", "revenues", "revenue_change"]].head()

In [3]:
f500_selection

rank  revenues  revenue_change
Walmart                      1    485873             0.8
State Grid                   2    315199            -4.4
Sinopec Group                3    267518            -9.1
China National Petroleum     4    262573           -12.3
Toyota Motor                 5    254694             7.7

### Reading CSV files with pandas

1. Use the `pandas.read_csv()` function to read the `f500.csv` CSV file as a pandas dataframe. Assign it to the variable name `f500`.
- Do not use the `index_col` parameter.
2. Use the code below to insert the NaN values into the `previous_rank` column: 

`f500.loc[f500["previous_rank"] == 0, "previous_rank"] = np.nan`

In [4]:
f500 = pd.read_csv("f500.csv")
f500.loc[f500["previous_rank"] == 0, "previous_rank"] = np.nan

In [5]:
f500.head()

company  rank  revenues  revenue_change  profits  assets  \
0                   Walmart     1    485873             0.8  13643.0  198825   
1                State Grid     2    315199            -4.4   9571.3  489838   
2             Sinopec Group     3    267518            -9.1   1257.9  310726   
3  China National Petroleum     4    262573           -12.3   1867.5  585619   
4              Toyota Motor     5    254694             7.7  16899.3  437575   

   profit_change                  ceo                  industry  \
0           -7.2  C. Douglas McMillon     General Merchandisers   
1           -6.2              Kou Wei                 Utilities   
2          -65.0            Wang Yupu        Petroleum Refining   
3          -73.7        Zhang Jianhua        Petroleum Refining   
4          -12.3          Akio Toyoda  Motor Vehicles and Parts   

                   sector  previous_rank country      hq_location  \
0               Retailing            1.0     USA  Bentonville, AR   
1                  Energy            2.0   China   Beijing, China   
2                  Energy            4.0   China   Beijing, China   
3                  Energy            3.0   China   Beijing, China   
4  Motor Vehicles & Parts            8.0   Japan    Toyota, Japan   

                        website  years_on_global_500_list  employees  \
0        http://www.walmart.com                        23    2300000   
1        http://www.sgcc.com.cn                        17     926067   
2        http://www.sinopec.com                        19     713288   
3        http://www.cnpc.com.cn                        17    1512048   
4  http://www.toyota-global.com                        23     364445   

   total_stockholder_equity  
0                     77798  
1                    209456  
2                    106523  
3                    301893  
4                    157210

### Using iloc to Select by Integer Position

Without using the `index_col` parameter:
- The `company` column is now included as a regular column, instead of being used for the index.
- The index labels are now integers starting from `0`.

_Note_: This is the more conventional way to read in a dataframe and will be used from here on out.

We can use _integer positions_ to select data using `DataFrame.iloc[]` and `Series.iloc[]`.
- **l**oc: **l**abel based selection
- **i**loc: **integer** position based selection

**Syntax**

`df.iloc[row_index, column_index]`

1. Select just the fifth row of the `f500` dataframe. Assign the result to `fifth_row`.

In [6]:
fifth_row = f500.iloc[4]

In [7]:
fifth_row

company                                     Toyota Motor
rank                                                   5
revenues                                          254694
revenue_change                                       7.7
profits                                          16899.3
assets                                            437575
profit_change                                      -12.3
ceo                                          Akio Toyoda
industry                        Motor Vehicles and Parts
sector                            Motor Vehicles & Parts
previous_rank                                          8
country                                            Japan
hq_location                                Toyota, Japan
website                     http://www.toyota-global.com
years_on_global_500_list                              23
employees                                         364445
total_stockholder_equity                          157210
Name: 4, dtype: object

2. Select the value in the first row of the `company` column. Assign the result to `company_value`.

In [8]:
company_value = f500["company"].iloc[0]

In [9]:
company_value

'Walmart'

### Using iloc to Select by Integer Position (Continued)

To select a whole column:

In [10]:
first_column = f500.iloc[:, 0]
first_column

0                             Walmart
1                          State Grid
2                       Sinopec Group
3            China National Petroleum
4                        Toyota Motor
                    ...              
495    Teva Pharmaceutical Industries
496          New China Life Insurance
497         Wm. Morrison Supermarkets
498                               TUI
499                        AutoNation
Name: company, Length: 500, dtype: object

To select a positional slice:

In [11]:
second_to_sixth_rows = f500[1:5]
second_to_sixth_rows

company  rank  revenues  revenue_change  profits  assets  \
1                State Grid     2    315199            -4.4   9571.3  489838   
2             Sinopec Group     3    267518            -9.1   1257.9  310726   
3  China National Petroleum     4    262573           -12.3   1867.5  585619   
4              Toyota Motor     5    254694             7.7  16899.3  437575   

   profit_change            ceo                  industry  \
1           -6.2        Kou Wei                 Utilities   
2          -65.0      Wang Yupu        Petroleum Refining   
3          -73.7  Zhang Jianhua        Petroleum Refining   
4          -12.3    Akio Toyoda  Motor Vehicles and Parts   

                   sector  previous_rank country     hq_location  \
1                  Energy            2.0   China  Beijing, China   
2                  Energy            4.0   China  Beijing, China   
3                  Energy            3.0   China  Beijing, China   
4  Motor Vehicles & Parts            8.0   Japan   Toyota, Japan   

                        website  years_on_global_500_list  employees  \
1        http://www.sgcc.com.cn                        17     926067   
2        http://www.sinopec.com                        19     713288   
3        http://www.cnpc.com.cn                        17    1512048   
4  http://www.toyota-global.com                        23     364445   

   total_stockholder_equity  
1                    209456  
2                    106523  
3                    301893  
4                    157210

- With `loc[]`, the ending slice **is** included.
- With `iloc[]`, the ending slice **is not** included.

| Select by integer position | Explicit Syntax | Shorthand Convention |
| :--- | :--- | :---|
| Single column from dataframe | `df.iloc[:,[3,5,6]]` | |
| List of columns from dataframe | `df.iloc[:,3]` | |
| Slice of columns from dataframe | `df.iloc[:,3:7]` | |
| Single row from dataframe | `df.iloc[20]` | |
| List of rows from dataframe | `df.iloc[[0,3,8]]` | |
| Slice of rows from dataframe | `df.iloc[3:5]` | `df[3:5]` |
| Single items from series | `s,iloc[8]` | `s[8]` |
| List of items from series | `s.iloc[[2,8,1]]` | `s[[2,8,1]]` |
| Slice of items from series | `s.iloc[5:10]` | `s[5:10]` |

1. Select the first three rows of the `f500` dataframe. Assign the result to `first_three_rows`.

In [12]:
first_three_rows = f500.iloc[:3]
first_three_rows

company  rank  revenues  revenue_change  profits  assets  \
0        Walmart     1    485873             0.8  13643.0  198825   
1     State Grid     2    315199            -4.4   9571.3  489838   
2  Sinopec Group     3    267518            -9.1   1257.9  310726   

   profit_change                  ceo               industry     sector  \
0           -7.2  C. Douglas McMillon  General Merchandisers  Retailing   
1           -6.2              Kou Wei              Utilities     Energy   
2          -65.0            Wang Yupu     Petroleum Refining     Energy   

   previous_rank country      hq_location                 website  \
0            1.0     USA  Bentonville, AR  http://www.walmart.com   
1            2.0   China   Beijing, China  http://www.sgcc.com.cn   
2            4.0   China   Beijing, China  http://www.sinopec.com   

   years_on_global_500_list  employees  total_stockholder_equity  
0                        23    2300000                     77798  
1                        17     926067                    209456  
2                        19     713288                    106523

2. Select the first and seventh rows and the first five columns of the `f500` dataframe. Assign the result to `first_seventh_row_slice`.

In [13]:
first_seventh_row_slice = f500.iloc[[0,6], :5]
first_seventh_row_slice

company  rank  revenues  revenue_change  profits
0            Walmart     1    485873             0.8  13643.0
6  Royal Dutch Shell     7    240033           -11.8   4575.0

### Using pandas Methods to Create Boolean Masks

Use `Series.isnull()` to view rows with null values in the `revenue_change` column:

In [14]:
rev_is_null = f500["revenue_change"].isnull()
rev_is_null.head()

0    False
1    False
2    False
3    False
4    False
Name: revenue_change, dtype: bool

In [15]:
rev_change_null = f500[rev_is_null]
rev_change_null[["company", "country", "sector"]]

company  country      sector
90                       Uniper  Germany      Energy
180  Hewlett Packard Enterprise      USA  Technology

1. Use the `Series.isnull()` method to select all rows from `f500` that have a null value for the `previous_rank` column. Select only the `company`, `rank`, and `previous_rank` columns. Assign the result to `null_previous_rank`.

In [16]:
null_previous_rank = f500[f500["previous_rank"].isnull()][["company", "rank", "previous_rank"]]
null_previous_rank

company  rank  previous_rank
48                   Legal & General Group    49            NaN
90                                  Uniper    91            NaN
123                      Dell Technologies   124            NaN
138                 Anbang Insurance Group   139            NaN
140                        Albertsons Cos.   141            NaN
180             Hewlett Packard Enterprise   181            NaN
267                           Hengli Group   268            NaN
271         Johnson Controls International   272            NaN
341                                  Chubb   342            NaN
375                 Charter Communications   376            NaN
376                              Medtronic   377            NaN
392                               Facebook   393            NaN
409                            Kraft Heinz   410            NaN
414                                  Nokia   415            NaN
431                          Standard Life   432            NaN
440            Brookfield Asset Management   441            NaN
441        Mercantil Servicios Financieros   442            NaN
446                            Samsung C&T   447            NaN
456                        Heraeus Holding   457            NaN
458                Yango Financial Holding   459            NaN
461                  Alibaba Group Holding   462            NaN
463                            Progressive   464            NaN
466                Country Garden Holdings   467            NaN
472                                 Altice   473            NaN
477                       Tencent Holdings   478            NaN
482                                  Aflac   483            NaN
484                  Suning Commerce Group   485            NaN
487                             Xiamen C&D   488            NaN
491                         Dollar General   492            NaN
493               Xiamen ITG Holding Group   494            NaN
494  Xinjiang Guanghui Industry Investment   495            NaN
495         Teva Pharmaceutical Industries   496            NaN
499                             AutoNation   500            NaN

### Working with Integer Labels

If we wanted to select the first company from our new `null_previous_rank` dataframe by _integer position_, we can use `DataFrame.iloc[]`:

In [17]:
first_null_prev_rank = null_previous_rank.iloc[0]
first_null_prev_rank

company          Legal & General Group
rank                                49
previous_rank                      NaN
Name: 48, dtype: object

If we wanted to select a row using `loc[]`, we'd have to use the integer label for the first row -- `48`.

1. Assign the first five rows of the `null_previous_rank` dataframe to the variable `top5_null_prev_rank` by choosing the correct method out of either `loc[]` or `iloc[]`.

In [19]:
null_previous_rank = f500[f500["previous_rank"].isnull()]
null_previous_rank

company  rank  revenues  revenue_change  \
48                   Legal & General Group    49    105235           442.3   
90                                  Uniper    91     74407             NaN   
123                      Dell Technologies   124     64806            18.1   
138                 Anbang Insurance Group   139     60800           124.0   
140                        Albertsons Cos.   141     59678             1.6   
180             Hewlett Packard Enterprise   181     50123             NaN   
267                           Hengli Group   268     37880            12.2   
271         Johnson Controls International   272     37674            -6.3   
341                                  Chubb   342     31469            65.7   
375                 Charter Communications   376     29003           197.3   
376                              Medtronic   377     28833            42.3   
392                               Facebook   393     27638            54.2   
409                            Kraft Heinz   410     26487            44.4   
414                                  Nokia   415     26113            73.4   
431                          Standard Life   432     25279            46.9   
440            Brookfield Asset Management   441     24411            22.6   
441        Mercantil Servicios Financieros   442     24403            50.3   
446                            Samsung C&T   447     24217           105.2   
456                        Heraeus Holding   457     23793            65.9   
458                Yango Financial Holding   459     23657            11.5   
461                  Alibaba Group Holding   462     23517            47.9   
463                            Progressive   464     23441            12.4   
466                Country Garden Holdings   467     23044            27.9   
472                                 Altice   473     22953            42.2   
477                       Tencent Holdings   478     22871            39.7   
482                                  Aflac   483     22559             8.1   
484                  Suning Commerce Group   485     22366             3.7   
487                             Xiamen C&D   488     22145             6.6   
491                         Dollar General   492     21987             7.9   
493               Xiamen ITG Holding Group   494     21930            34.3   
494  Xinjiang Guanghui Industry Investment   495     21919            31.1   
495         Teva Pharmaceutical Industries   496     21903            11.5   
499                             AutoNation   500     21609             3.6   

     profits  assets  profit_change                      ceo  \
48    1697.9  577954            3.4             Nigel Wilson   
90   -3557.5   51541            NaN            Klaus Schafer   
123  -1672.0  118206            NaN          Michael S. Dell   
138   3883.9  430040            0.9               Wu Xiaohui   
140   -373.3   23755            NaN         Robert G. Miller   
180   3161.0   79679            NaN      Margaret C. Whitman   
267    821.7   14838           43.7             Chen Jianhua   
271   -868.0   63253         -155.5       Alex A. Molinaroli   
341   4135.0  159786           45.9        Evan G. Greenberg   
375   3522.0  149067            NaN       Thomas M. Rutledge   
376   3538.0   99782           32.3           Omar S. Ishrak   
392  10217.0   64961          177.0          Mark Zuckerberg   
409   3632.0  120480          472.9            Bernardo Hees   
414   -847.1   47354         -131.0              Rajeev Suri   
431    496.7  235324          -77.2             Keith Skeoch   
440   1651.0  159826          -29.5           J. Bruce Flatt   
441   2004.2  148659          -10.5    Gustavo J. Vollmer A.   
446     92.5   36816          -96.2             Chi-Hun Choi   
456      NaN    4982            NaN              Jan Rinnert   
458    159.2   24231          -59.0             Lin Tengjiao   
461   6489.5   73538          -42.2   

In [20]:
top5_null_prev_rank = null_previous_rank.iloc[:5]
top5_null_prev_rank

company  rank  revenues  revenue_change  profits  assets  \
48    Legal & General Group    49    105235           442.3   1697.9  577954   
90                   Uniper    91     74407             NaN  -3557.5   51541   
123       Dell Technologies   124     64806            18.1  -1672.0  118206   
138  Anbang Insurance Group   139     60800           124.0   3883.9  430040   
140         Albertsons Cos.   141     59678             1.6   -373.3   23755   

     profit_change               ceo                          industry  \
48             3.4      Nigel Wilson   Insurance: Life, Health (stock)   
90             NaN     Klaus Schafer                            Energy   
123            NaN   Michael S. Dell       Computers, Office Equipment   
138            0.9        Wu Xiaohui  Insurance: Life, Health (Mutual)   
140            NaN  Robert G. Miller              Food and Drug Stores   

                 sector  previous_rank  country          hq_location  \
48           Financials            NaN  Britain      London, Britain   
90               Energy            NaN  Germany  Dusseldorf, Germany   
123          Technology            NaN      USA       Round Rock, TX   
138          Financials            NaN    China       Beijing, China   
140  Food & Drug Stores            NaN      USA            Boise, ID   

                                 website  years_on_global_500_list  employees  \
48   http://www.legalandgeneralgroup.com                        17       8939   
90              http://www.uniper.energy                         1      12890   
123      http://www.delltechnologies.com                        17     138000   
138           http://www.anbanggroup.com                         1      40707   
140            http://www.albertsons.com                        13     273000   

     total_stockholder_equity  
48                       8579  
90                      12889  
123                     13243  
138                     20372  
140                      1371

### Pandas Index Alignment

Now that we've identified the rows with null values in the `previous_rank` column, let's use the `Series.notnull()` method to exclude them from the next part of the analysis.


In [21]:
previously_ranked = f500[f500["previous_rank"].notnull()]

Then create a `rank_change` column by subtracting the `rank` column from the `previous_rank` column:

In [22]:
rank_change = previously_ranked["previous_rank"] - previously_ranked["rank"]
rank_change.shape

(467,)

In [25]:
rank_change.tail(3)

496   -70.0
497   -61.0
498   -32.0
dtype: float64

Since the `rank_change` has 467 rows and the last integer index label is 498, we know that our index labels no longer align wiht the integer positions.

Suppose now we decided to add the `rank_change` series to the `f500` dataframe as a new column. Its index labels no longer match the index labels in `f500`, so how can this be done?

Another powerful aspect of pandas is that almost every operation will align on the index labels. The pandas library will align on index at every opportunity, no matter if our index labels are strings or integers -- this makes working with data from different sources or working with data when we have removed, added, or reordered rows much easier than it would be otherwise.

1. Use the `Series.notnull()` method to select all rows from `f500` that have a non-null value for the `previous_rank` column. Assign the result to `previously_ranked`.

In [26]:
previously_ranked = f500[f500["previous_rank"].notnull()]

2. From the `previously_ranked` dataframe, subtract the `rank` column from the `previous_rank` column. Assign the result to `rank_change`.

In [27]:
rank_change = previously_ranked["previous_rank"] - previously_ranked["rank"]

3. Assign the values in the `rank_change` to a new column in the `f500` dataframe, `"rank_change"`.

In [28]:
f500["rank_change"] = rank_change

In [29]:
f500.head()

company  rank  revenues  revenue_change  profits  assets  \
0                   Walmart     1    485873             0.8  13643.0  198825   
1                State Grid     2    315199            -4.4   9571.3  489838   
2             Sinopec Group     3    267518            -9.1   1257.9  310726   
3  China National Petroleum     4    262573           -12.3   1867.5  585619   
4              Toyota Motor     5    254694             7.7  16899.3  437575   

   profit_change                  ceo                  industry  \
0           -7.2  C. Douglas McMillon     General Merchandisers   
1           -6.2              Kou Wei                 Utilities   
2          -65.0            Wang Yupu        Petroleum Refining   
3          -73.7        Zhang Jianhua        Petroleum Refining   
4          -12.3          Akio Toyoda  Motor Vehicles and Parts   

                   sector  previous_rank country      hq_location  \
0               Retailing            1.0     USA  Bentonville, AR   
1                  Energy            2.0   China   Beijing, China   
2                  Energy            4.0   China   Beijing, China   
3                  Energy            3.0   China   Beijing, China   
4  Motor Vehicles & Parts            8.0   Japan    Toyota, Japan   

                        website  years_on_global_500_list  employees  \
0        http://www.walmart.com                        23    2300000   
1        http://www.sgcc.com.cn                        17     926067   
2        http://www.sinopec.com                        19     713288   
3        http://www.cnpc.com.cn                        17    1512048   
4  http://www.toyota-global.com                        23     364445   

   total_stockholder_equity  rank_change  
0                     77798          0.0  
1                    209456          0.0  
2                    106523          1.0  
3                    301893         -1.0  
4                    157210          3.0

### Using Boolean Operators

In [30]:
cols = ["company", "revenues", "country"]
f500_sel = f500[cols].head()
f500_sel

company  revenues country
0                   Walmart    485873     USA
1                State Grid    315199   China
2             Sinopec Group    267518   China
3  China National Petroleum    262573   China
4              Toyota Motor    254694   Japan

In [31]:
over_265 = f500_sel["revenues"] > 265000
china = f500_sel["country"] == "China"
combined = over_265 & china

In [32]:
combined

0    False
1     True
2     True
3    False
4    False
dtype: bool

In [33]:
final_cols = ["company", "revenues"]
result = f500_sel.loc[combined, final_cols]
result

company  revenues
1     State Grid    315199
2  Sinopec Group    267518

1. Select all companies with revenues over 100 billion and negative profits from the `f500` dataframe. The result should include all columns.
- Create a boolean array that selects the companies with revenues greater than 100 billion. Assign the result to `large_revenue`.
- Create a boolean array that selects the companies with profits less than 0. Assign the result to `negative_profits`.
- Combine `large_revenue` and `negative_profits`. Assign the result to `combined`.
- Use `combined` to filter `f500`. Assign the result to `big_rev_neg_profit`.

In [34]:
large_revenue = f500["revenues"] > 100000
negative_profits = f500["profits"] < 0
combined = large_revenue & negative_profits
big_rev_neg_profit = f500[combined]
big_rev_neg_profit

company  rank  revenues  revenue_change  profits   assets  \
32  Japan Post Holdings    33    122990             3.6   -267.4  2631385   
44              Chevron    45    107567           -18.0   -497.0   260078   

    profit_change               ceo                         industry  \
32         -107.5  Masatsugu Nagato  Insurance: Life, Health (stock)   
44         -110.8    John S. Watson               Petroleum Refining   

        sector  previous_rank country    hq_location                  website  \
32  Financials           37.0   Japan   Tokyo, Japan  http://www.japanpost.jp   
44      Energy           31.0     USA  San Ramon, CA   http://www.chevron.com   

    years_on_global_500_list  employees  total_stockholder_equity  rank_change  
32                        21     248384                     91532          4.0  
44                        23      55200                    145556        -14.0

### Using Boolean Operators Continued

A more concise way of writing the above code is to remove the intermediate variables `large_revenue` and `negative_profits`:

`combined = (f500["revenues"] > 100000) & (f500["profits"] < 0)`

OR

`big_rev_neg_profit = f500[(f500["revenues"] > 100000) & (f500["profits"] < 0)]`

1. Select all rows for companies whose `country` value is either Brazil or Venezuela. Assign the result to `brazil_venezuela`.

In [36]:
brazil_venezuela = f500[(f500["country"] == "Brazil") | (f500["country"] == "Venezuela")]
brazil_venezuela

company  rank  revenues  revenue_change  profits  \
74                         Petrobras    75     81405           -16.3  -4838.0   
112            Itau Unibanco Holding   113     66876            21.4   6666.4   
150                  Banco do Brasil   151     58093           -13.4   2013.8   
153                   Banco Bradesco   154     57443            31.3   5127.9   
190                              JBS   191     48825            -0.1    107.7   
369                             Vale   370     29363            14.7   3982.0   
441  Mercantil Servicios Financieros   442     24403            50.3   2004.2   
486                Ultrapar Holdings   487     22167            -2.3    447.5   

     assets  profit_change                        ceo  \
74   246983            NaN       Pedro Pullen Parente   
112  415972          -13.7    Candido Botelho Bracher   
150  426416          -52.3   Paulo Rogerio Caffarelli   
153  366418           -5.7  Luiz Carlos Trabuco Cappi   
190   31605          -92.3    Wesley Mendonca Batista   
369   99014            NaN          Fabio Schvartsman   
441  148659          -10.5      Gustavo J. Vollmer A.   
486    7426           -0.8            Thilo Mannhardt   

                          industry                     sector  previous_rank  \
74              Petroleum Refining                     Energy           58.0   
112  Banks: Commercial and Savings                 Financials          159.0   
150  Banks: Commercial and Savings                 Financials          115.0   
153  Banks: Commercial and Savings                 Financials          209.0   
190                Food Production  Food, Beverages & Tobacco          185.0   
369   Mining, Crude-Oil Production                     Energy          417.0   
441  Banks: Commercial and Savings                 Financials            NaN   
486                         Energy                     Energy          474.0   

       country             hq_location                        website  \
74      Brazil  Rio de Janeiro, Brazil    http://www.petrobras.com.br   
112     Brazil       Sao Paulo, Brazil         http://www.itau.com.br   
150     Brazil        Brasilia, Brazil           http://www.bb.com.br   
153     Brazil          Osasco, Brazil     http://www.bradesco.com.br   
190     Brazil       Sao Paulo, Brazil  http://jbss.infoinvest.com.br   
369     Brazil  Rio de Janeiro, Brazil            http://www.vale.com   
441  Venezuela      Caracas, Venezuela             http://www.msf.com   
486     Brazil       Sao Paulo, Brazil        http://www.ultra.com.br   

     years_on_global_500_list  employees  total_stockholder_equity  \
74                         23      68829                     76779   
112                         4      94779                     37680   
150                        23     100622                     26551   
153                        21      94541                     32369   
190                         8     237061                      7307   
369                        11      73062                     39042   
441                         1       8370                      7550   
486                         8      15173                      2621   

     rank_change  
74         -17.0  
112         46.0  
150        -36.0  
153         55.0  
190         -6.0  
369         47.0  
441          NaN  
486        -13.0

2. Select the first five companies in the Technology sector for which the country is **not** the USA from the `f500` dataframe. Assign the result to `tech_outside_usa`.

In [37]:
tech_outside_usa = f500[(f500["sector"] == "Technology") & (f500["country"] != "USA")].head()
tech_outside_usa

company  rank  revenues  revenue_change  profits  \
14           Samsung Electronics    15    173957            -2.0  19316.5   
26    Hon Hai Precision Industry    27    135129            -4.3   4608.8   
70                       Hitachi    71     84558             1.2   2134.3   
82   Huawei Investment & Holding    83     78511            24.9   5579.4   
104                         Sony   105     70170             3.9    676.4   

     assets  profit_change                   ceo  \
14   217104           16.8          Oh-Hyun Kwon   
26    80436           -0.4             Terry Gou   
70    86742           48.8  Toshiaki Higashihara   
82    63837           -5.0          Ren Zhengfei   
104  158519          -45.1           Kazuo Hirai   

                                       industry      sector  previous_rank  \
14               Electronics, Electrical Equip.  Technology           13.0   
26               Electronics, Electrical Equip.  Technology           25.0   
70               Electronics, Electrical Equip.  Technology           79.0   
82   Network and Other Communications Equipment  Technology          129.0   
104              Electronics, Electrical Equip.  Technology          113.0   

         country              hq_location                 website  \
14   South Korea       Suwon, South Korea  http://www.samsung.com   
26        Taiwan  New Taipei City, Taiwan  http://www.foxconn.com   
70         Japan             Tokyo, Japan  http://www.hitachi.com   
82         China          Shenzhen, China   http://www.huawei.com   
104        Japan             Tokyo, Japan     http://www.sony.net   

     years_on_global_500_list  employees  total_stockholder_equity  \
14                         23     325000                    154376   
26                         13     726772                     33476   
70                         23     303887                     26632   
82                          8     180000                     20159   
104                        23     128400                     22415   

     rank_change  
14          -2.0  
26          -2.0  
70           8.0  
82          46.0  
104          8.0

### Sorting Values

Select the company that employs the most people in China by first selecting all of the rows where the `country` column equals `China`:

In [38]:
selected_rows = f500[f500["country"] == "China"]

Use the `DataFrame.sort_values()` method to sort the rows on the `employees` column by passing the column name to the method:

In [39]:
sorted_rows = selected_rows.sort_values("employees")
sorted_rows[["company", "country", "employees"]].head()

company country  employees
204                         Noble Group   China       1000
458             Yango Financial Holding   China      10234
438  China National Aviation Fuel Group   China      11739
128                         Tewoo Group   China      17353
182            Amer International Group   China      17852

By default, the `sort_values()` method will sort the rows in _ascending_ order. To sort the rows in _descending_ order instead, we can set the `ascending` parameter to `False`.

In [40]:
sorted_rows = selected_rows.sort_values("employees", ascending=False)
sorted_rows[["company", "country", "employees"]].head()

company country  employees
3      China National Petroleum   China    1512048
118            China Post Group   China     941211
1                    State Grid   China     926067
2                 Sinopec Group   China     713288
37   Agricultural Bank of China   China     501368

1. Find the company headquartered in Japan with the largest number of employees.
- Select only the rows that have a country name equal to `Japan`.
- Use `DataFrame.sort_values()` to sort those rows _by_ the `employees` column in _descending_ order.
- Use `DataFrame.iloc[]` to select the first row from the sorted dataframe.
- Extract the company name from the index label `company` from the first row. Assign the result to `top_japanese_employer`.

In [41]:
selected_rows = f500[f500["country"] == "Japan"]
selected_rows

company  rank  revenues  revenue_change  profits  \
4                       Toyota Motor     5    254694             7.7  16899.3   
28                       Honda Motor    29    129198             6.2   5690.3   
32               Japan Post Holdings    33    122990             3.6   -267.4   
43                      Nissan Motor    44    108164             6.5   6123.4   
49      Nippon Telegraph & Telephone    50    105128             9.4   7384.4   
70                           Hitachi    71     84558             1.2   2134.3   
71                    SoftBank Group    72     82892             8.4  13163.4   
86                              AEON    87     75772            11.8    103.9   
104                             Sony   105     70170             3.9    676.4   
109                        Panasonic   110     67775             6.7   1378.4   
110            Nippon Life Insurance   111     67388             0.4   2786.9   
115                         Marubeni   116     65792             8.2   1433.7   
126                    JXTG Holdings   127     63629             1.8   1477.3   
141           Dai-ichi Life Holdings   142     59590            -2.5   2134.5   
144                       Mitsubishi   145     59303             2.8   4063.5   
163   Mitsubishi UFJ Financial Group   164     55185            15.9   8550.1   
166               Seven & I Holdings   167     53858             7.5    892.9   
184             Tokyo Electric Power   185     49446            -2.2   1225.7   
187   MS&AD Insurance Group Holdings   188     49239            17.9   1942.2   
192            Tokio Marine Holdings   193     48292            26.6   2527.4   
201  Sumitomo Mitsui Financial Group   202     47375            19.2   6520.5   
214                           Itochu   215     44654             5.5   3250.6   
218                             KDDI   219     43822            17.8   5045.1   
227    Nippon Steel & Sumitomo Metal   228     42757             4.6   1208.5   
235                            Denso   236     41781            10.9   2377.6   
236                          Fujitsu   237     41620             5.4    816.7   
241          Sumitomo Life Insurance   242     40921            31.6    517.5   
248                           Mitsui   249     40275             1.6   2825.3   
261              Mitsubishi Electric   262     39119             6.9   1942.6   
281                         Sumitomo   282     36888            10.4   1577.1   
293      Mitsubishi Heavy Industries   294     36122             7.2    809.6   
296      Meiji Yasuda Life Insurance   297     35767             0.4   2064.8   
323                      Aisin Seiki   324     32879            21.6   1168.9   
329             Daiwa House Industry   330     32421            21.9   1861.5   
339                   Sompo Holdings   340     31559            16.4   1535.7   
346                            Canon   347     31271            -0.4   1385.0   
348     Mitsubishi Chemical Holdings   349     31158             5.6   1442.1   
351                           Subaru   352     30696            14.0   2605.8   
352                      Bridgestone   353     30678            -2.0   2441.3   
355                     JFE Holdings   356     30539             6.8    627.0   
356           Mizuho Financial Group   357     30390            13.5   5570.1   
366                      Mazda Motor   367     29665             4.5    865.5   
372                     Suzuki Motor   373     29252            10.4   1476.2   
380                 Medipal Holdings   381     28277            12.1    267.7   
388            Kansai Electric Power   389     27792             2.8   1299.3   
407               East Japan Railway   408     26587            11.3   2565.0   
419     Sumitomo Electric Industries   420     25975             6.3    992.7   
422                   Idemitsu Kosan   423     25888            -1.0    813.7   
436                              NEC   437     24596        

In [43]:
sorted_rows = selected_rows.sort_values("employees", ascending=False)
sorted_rows

company  rank  revenues  revenue_change  profits  \
4                       Toyota Motor     5    254694             7.7  16899.3   
70                           Hitachi    71     84558             1.2   2134.3   
49      Nippon Telegraph & Telephone    50    105128             9.4   7384.4   
86                              AEON    87     75772            11.8    103.9   
109                        Panasonic   110     67775             6.7   1378.4   
32               Japan Post Holdings    33    122990             3.6   -267.4   
419     Sumitomo Electric Industries   420     25975             6.3    992.7   
28                       Honda Motor    29    129198             6.2   5690.3   
346                            Canon   347     31271            -0.4   1385.0   
236                          Fujitsu   237     41620             5.4    816.7   
235                            Denso   236     41781            10.9   2377.6   
352                      Bridgestone   353     30678            -2.0   2441.3   
261              Mitsubishi Electric   262     39119             6.9   1942.6   
43                      Nissan Motor    44    108164             6.5   6123.4   
104                             Sony   105     70170             3.9    676.4   
163   Mitsubishi UFJ Financial Group   164     55185            15.9   8550.1   
214                           Itochu   215     44654             5.5   3250.6   
436                              NEC   437     24596             4.5    252.0   
227    Nippon Steel & Sumitomo Metal   228     42757             4.6   1208.5   
323                      Aisin Seiki   324     32879            21.6   1168.9   
407               East Japan Railway   408     26587            11.3   2565.0   
110            Nippon Life Insurance   111     67388             0.4   2786.9   
293      Mitsubishi Heavy Industries   294     36122             7.2    809.6   
201  Sumitomo Mitsui Financial Group   202     47375            19.2   6520.5   
144                       Mitsubishi   145     59303             2.8   4063.5   
281                         Sumitomo   282     36888            10.4   1577.1   
348     Mitsubishi Chemical Holdings   349     31158             5.6   1442.1   
71                    SoftBank Group    72     82892             8.4  13163.4   
372                     Suzuki Motor   373     29252            10.4   1476.2   
141           Dai-ichi Life Holdings   142     59590            -2.5   2134.5   
329             Daiwa House Industry   330     32421            21.9   1861.5   
355                     JFE Holdings   356     30539             6.8    627.0   
356           Mizuho Financial Group   357     30390            13.5   5570.1   
166               Seven & I Holdings   167     53858             7.5    892.9   
366                      Mazda Motor   367     29665             4.5    865.5   
339                   Sompo Holdings   340     31559            16.4   1535.7   
248                           Mitsui   249     40275             1.6   2825.3   
241          Sumitomo Life Insurance   242     40921            31.6    517.5   
184             Tokyo Electric Power   185     49446            -2.2   1225.7   
296      Meiji Yasuda Life Insurance   297     35767             0.4   2064.8   
187   MS&AD Insurance Group Holdings   188     49239            17.9   1942.2   
115                         Marubeni   116     65792             8.2   1433.7   
192            Tokio Marine Holdings   193     48292            26.6   2527.4   
351                           Subaru   352     30696            14.0   2605.8   
218                             KDDI   219     43822            17.8   5045.1   
388            Kansai Electric Power   389     27792             2.8   1299.3   
450             Chubu Electric Power   451     24028             1.1   1058.2   
126                    JXTG Holdings   127     63629             1.8   1477.3   
380                 Medipal Holdings   381     28277        

In [45]:
top_japanese_employer = sorted_rows.iloc[0]["company"]
top_japanese_employer

'Toyota Motor'

### Using Loops with pandas

Loops for the purpose of aggregation:

If we wanted to calcualte the average revenue for each country in the data set, our process may look like this:
- Identify each unique country in the data set using the `Series.unique()` method.
- For each country:
    - Select only the rows corresponding to that country.
    - Calculate the average revenue for those rows.

In [46]:
# Create an empty dictionary to store the results
avg_rev_by_country = {}

# Create an array of unique countries
countries = f500["country"].unique()

# Use a for loop to iterate over the countries
for c in countries:
    # Use boolean comparison to select only rows that
    # correspond to a specific country
    selected_rows = f500[f500["country"] == c]
    # Calculate the mean average revenue for just those rows
    mean = selected_rows["revenues"].mean()
    # Assign the mean value to the dictionary, using the
    # country name as the key
    avg_rev_by_country[c] = mean

In [47]:
avg_rev_by_country

{'USA': 64218.371212121216,
 'China': 55397.880733944956,
 'Japan': 53164.03921568627,
 'Germany': 63915.0,
 'Netherlands': 61708.92857142857,
 'Britain': 51588.708333333336,
 'South Korea': 49725.6,
 'Switzerland': 51353.57142857143,
 'France': 55231.793103448275,
 'Taiwan': 46364.666666666664,
 'Singapore': 54454.333333333336,
 'Italy': 51899.57142857143,
 'Russia': 65247.75,
 'Spain': 40600.666666666664,
 'Brazil': 52024.57142857143,
 'Mexico': 54987.5,
 'Luxembourg': 56791.0,
 'India': 39993.0,
 'Malaysia': 49479.0,
 'Thailand': 48719.0,
 'Australia': 33688.71428571428,
 'Belgium': 45905.0,
 'Norway': 45873.0,
 'Canada': 31848.0,
 'Ireland': 32819.5,
 'Indonesia': 36487.0,
 'Denmark': 35464.0,
 'Saudi Arabia': 35421.0,
 'Sweden': 27963.666666666668,
 'Finland': 26113.0,
 'Venezuela': 24403.0,
 'Turkey': 23456.0,
 'U.A.E': 22799.0,
 'Israel': 21903.0}

1. Create an empty dictionary, `top_employer_by_country` to store the results of the exercise.

In [48]:
top_employer_by_country = {}

2. Use the `Series.unique()` method to create an array of unique values from the `country` column.

In [49]:
countries = f500["country"].unique()
countries

array(['USA', 'China', 'Japan', 'Germany', 'Netherlands', 'Britain',
       'South Korea', 'Switzerland', 'France', 'Taiwan', 'Singapore',
       'Italy', 'Russia', 'Spain', 'Brazil', 'Mexico', 'Luxembourg',
       'India', 'Malaysia', 'Thailand', 'Australia', 'Belgium', 'Norway',
       'Canada', 'Ireland', 'Indonesia', 'Denmark', 'Saudi Arabia',
       'Sweden', 'Finland', 'Venezuela', 'Turkey', 'U.A.E', 'Israel'],
      dtype=object)

3. Use a for loop to iterate over the array unqiue countries. In each iteration:
- Select only the rows that have a country name equal to the current iteration.
- Use `DataFrame.sort_values()` to sort those rows _by_ the `employees` column in _descending_ order.
- Select the first row from the sorted dataframe.
- Extract the company name from the index label `company` from the first row.
- Assign the results to the `top_employer_by_country` dictionary, using the country name as the key, and the compnay name as the value.

In [51]:
for c in countries:
    selected_rows = f500[f500["country"] == c]
    sorted_rows = selected_rows.sort_values("employees", ascending=False)
    top_employer = sorted_rows.iloc[0]
    employer_name = top_employer["company"]
    top_employer_by_country[c] = employer_name

In [52]:
top_employer_by_country

{'USA': 'Walmart',
 'China': 'China National Petroleum',
 'Japan': 'Toyota Motor',
 'Germany': 'Volkswagen',
 'Netherlands': 'EXOR Group',
 'Britain': 'Compass Group',
 'South Korea': 'Samsung Electronics',
 'Switzerland': 'Nestle',
 'France': 'Sodexo',
 'Taiwan': 'Hon Hai Precision Industry',
 'Singapore': 'Flex',
 'Italy': 'Poste Italiane',
 'Russia': 'Gazprom',
 'Spain': 'Banco Santander',
 'Brazil': 'JBS',
 'Mexico': 'America Movil',
 'Luxembourg': 'ArcelorMittal',
 'India': 'State Bank of India',
 'Malaysia': 'Petronas',
 'Thailand': 'PTT',
 'Australia': 'Wesfarmers',
 'Belgium': 'Anheuser-Busch InBev',
 'Norway': 'Statoil',
 'Canada': 'George Weston',
 'Ireland': 'Accenture',
 'Indonesia': 'Pertamina',
 'Denmark': 'Maersk Group',
 'Saudi Arabia': 'SABIC',
 'Sweden': 'H & M Hennes & Mauritz',
 'Finland': 'Nokia',
 'Venezuela': 'Mercantil Servicios Financieros',
 'Turkey': 'Koc Holding',
 'U.A.E': 'Emirates Group',
 'Israel': 'Teva Pharmaceutical Industries'}

### Challenge: Calculating Return on Assets by Country

Return on assets (ROA) is a business-specific metric which indicates a company's ability to make profit using their available assets.

`return on assets = profit / assets`

1. Create a new column `roa` in the `f500` dataframe, containing the return on assets metric for each company.

In [53]:
f500["roa"] = f500["profits"] / f500["assets"]
f500.head()

company  rank  revenues  revenue_change  profits  assets  \
0                   Walmart     1    485873             0.8  13643.0  198825   
1                State Grid     2    315199            -4.4   9571.3  489838   
2             Sinopec Group     3    267518            -9.1   1257.9  310726   
3  China National Petroleum     4    262573           -12.3   1867.5  585619   
4              Toyota Motor     5    254694             7.7  16899.3  437575   

   profit_change                  ceo                  industry  \
0           -7.2  C. Douglas McMillon     General Merchandisers   
1           -6.2              Kou Wei                 Utilities   
2          -65.0            Wang Yupu        Petroleum Refining   
3          -73.7        Zhang Jianhua        Petroleum Refining   
4          -12.3          Akio Toyoda  Motor Vehicles and Parts   

                   sector  previous_rank country      hq_location  \
0               Retailing            1.0     USA  Bentonville, AR   
1                  Energy            2.0   China   Beijing, China   
2                  Energy            4.0   China   Beijing, China   
3                  Energy            3.0   China   Beijing, China   
4  Motor Vehicles & Parts            8.0   Japan    Toyota, Japan   

                        website  years_on_global_500_list  employees  \
0        http://www.walmart.com                        23    2300000   
1        http://www.sgcc.com.cn                        17     926067   
2        http://www.sinopec.com                        19     713288   
3        http://www.cnpc.com.cn                        17    1512048   
4  http://www.toyota-global.com                        23     364445   

   total_stockholder_equity  rank_change       roa  
0                     77798          0.0  0.068618  
1                    209456          0.0  0.019540  
2                    106523          1.0  0.004048  
3                    301893         -1.0  0.003189  
4                    157210          3.0  0.038620

2. Aggregate the data by the `sector` column, and create a dictionary `top_roa_by_sector`, with:
- Dictionary keys with the sector name.
- Dictionary values with the company name with the highest ROA value from that sector.

In [54]:
top_roa_by_sector = {}

In [55]:
for sector in f500["sector"].unique():
    is_sector = f500["sector"] == sector
    sector_companies = f500.loc[is_sector]
    top_company = sector_companies.sort_values("roa", ascending=False).iloc[0]
    company_name = top_company["company"]
    top_roa_by_sector[sector] = company_name

In [56]:
top_roa_by_sector

{'Retailing': 'H & M Hennes & Mauritz',
 'Energy': 'National Grid',
 'Motor Vehicles & Parts': 'Subaru',
 'Financials': 'Berkshire Hathaway',
 'Technology': 'Accenture',
 'Wholesalers': 'McKesson',
 'Health Care': 'Gilead Sciences',
 'Telecommunications': 'KDDI',
 'Engineering & Construction': 'Pacific Construction Group',
 'Industrials': '3M',
 'Food & Drug Stores': 'Publix Super Markets',
 'Aerospace & Defense': 'Lockheed Martin',
 'Food, Beverages & Tobacco': 'Philip Morris International',
 'Household Products': 'Unilever',
 'Transportation': 'Delta Air Lines',
 'Materials': 'CRH',
 'Chemicals': 'LyondellBasell Industries',
 'Media': 'Disney',
 'Apparel': 'Nike',
 'Hotels, Restaurants & Leisure': 'McDonald’s',
 'Business Services': 'Adecco Group'}